### PR TITLE
also include patch files in --new-pr and --update-pr

### PR DESCRIPTION
**Pull request by Caylo**
building on #1751 
cfr #1674

### CODE REVIEW

File: easybuild/tools/github.py [hunk 1]
```diff
@@ -595,8 +595,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                if patch_file in ec['patches']:
+                    soft_name = ec.name
+                    break
+
+            # FIXME: try harder if software name wasn't found yet?
+
+            if soft_name:
+                patch_specs.append((patch_path, soft_name))
+            else:
+                raise EasyBuildError("Failed to determine software name to which patch file %s relates", patch_path)
```
Comment (boegel) [inline easybuild/tools/github.py]: @Caylo please hoist this blob of code that determines `patch_specs` into a separate function, it'll help make things clearer since it's rather big (and will get bigger if the FIXME is handled)

File: easybuild/tools/github.py [hunk 2]
```diff
@@ -595,8 +595,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            soft_name = None
+            patch_file = os.path.basename(patch_path)
+
+            # consider patch lists of easyconfigs being provided
+            for ec in file_info['ecs']:
+                if patch_file in ec['patches']:
+                    soft_name = ec.name
+                    break
+
+            # FIXME: try harder if software name wasn't found yet?
```
Comment (boegel) [inline easybuild/tools/github.py]: @Caylo hmm, ideally, this should be tackled too...

As it is, only the easyconfig files that are being tweaked are being considered when trying to figure out which software the patch belongs too, but this may not be enough in practice

For example:

```
eb --new-pr foo.eb
# oops, forgot patch
eb --update-pr 1234 foo.patch
```

So, I think we should fall back to scanning all easyconfigs, and check which patches they require... This may take a while, but it's the best we can do imho (we can't rely on the patch name).

File: easybuild/framework/easyconfig/easyconfig.py [hunk 3]
```diff
@@ -1359,6 +1359,39 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
+        try:
+            mkdir(os.path.dirname(full_target_path), parents=True)
+            shutil.copy2(path, full_target_path)
+            _log.info("%s copied to %s", path, full_target_path)
+        except OSError as err:
+            raise EasyBuildError("Failed to copy %s to %s: %s", path, target_path, err)
+    else:
+        raise EasyBuildError("Subdirectory %s not found in %s", subdir, target_dir)
+
+    return target_path
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: this should be `return full_target_path`?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 4]
```diff
@@ -1373,42 +1406,36 @@ def copy_easyconfigs(paths, target_dir):
...
 
-                letter = name.lower()[0]
-                if letter not in a_to_z:
-                    raise EasyBuildError("Don't know which letter subdir to use for %s", name)
+            target_path = _copy_file_to_easyconfigs_dir(path, target_dir, soft_name, ec_filename)
+            full_target_path = os.path.join(target_dir, target_path)
 
-                target_path = os.path.join(subdir, letter, name, ec_filename)
-                _log.debug("Target path for %s: %s", path, target_path)
+            file_info['new'].append(not os.path.exists(full_target_path))
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: hmm, the `exists` check will always return `True` here, since it was just copied?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 5]
```diff
@@ -1373,42 +1406,36 @@ def copy_easyconfigs(paths, target_dir):
...
+    """
+    Copy patch files to specified directory, in the 'right' location according to the software name they relate to.
 
-    return file_info
+    @param patch_specs: list of tuples with patch file location and name of software they are for
+    @param target_dir: target directory
+    """
+    patched_files = []
+    for patch_path, soft_name in patch_specs:
+        patched_files.append(_copy_file_to_easyconfigs_dir(patch_path, target_dir, soft_name, os.path.basename(patch_path)))
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: line too long?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 6]
```diff
@@ -1365,6 +1365,34 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
 
+def det_easyconfig_location(path, target_dir, soft_name, target_file):
+    """
+    Determine path to easyconfigs directory for specified software name, using specified target file name.
+
+    @param path: path of file to copy
+    @param target_dir: (parent) target directory, should contain easybuild/easyconfigs subdirectory
+    @param soft_name: software name (to determine location to copy to)
+    @param target_file: target file name
+    @return: full path to the right location
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: `s/@param/:param/`

File: easybuild/framework/easyconfig/easyconfig.py [hunk 7]
```diff
@@ -1379,41 +1407,59 @@ def copy_easyconfigs(paths, target_dir):
...
+    @param patch_specs: list of tuples with patch file location and name of software they are for
+    @param target_dir: target directory
+    """
+    patched_files = {
+        'files': [],
+        'paths_in_repo': [],
+        'new': [],
+    }
+    for patch_path, soft_name in patch_specs:
+        target_path = det_easyconfig_location(patch_path, target_dir, soft_name, os.path.basename(patch_path))
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: since `det_easyconfig_location` is also used for patch files, we should rename it to `det_location_for`?

File: easybuild/framework/easyconfig/easyconfig.py [hunk 8]
```diff
@@ -1379,41 +1407,59 @@ def copy_easyconfigs(paths, target_dir):
...
+        patched_files['new'].append(not os.path.exists(target_path))
+
+        try:
+            mkdir(os.path.dirname(target_path), parents=True)
+            shutil.copy2(patch_path, target_path)
+            _log.info("%s copied to %s", patch_path, target_path)
+
+        except OSError as err:
+            raise EasyBuildError("Failed to copy %s to %s: %s", patch_path, target_path, err)
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: this whole block is copy-pasted, let's make a `copy_file` function in `filetools` for this?

I'm sure it can be used in other places as well (but that's beyond the scope of this PR)...

File: easybuild/framework/easyconfig/easyconfig.py [hunk 9]
```diff
@@ -1379,41 +1407,59 @@ def copy_easyconfigs(paths, target_dir):
...
+
+        patched_files['files'].append(os.path.splitext(os.path.basename(patch_path))[0])
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: hmm, what's the purpose of this `files` entry?

I'm fully aware that I may have been the one who added this, but it's no longer clear to me what this is about... ;-)

File: easybuild/tools/github.py [hunk 10]
```diff
@@ -617,14 +617,53 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                patch_specs.append((patch_path, soft_name))
+            else:
+                # fall back on scanning all eb files for patches
+                print "Matching easyconfig not found: determining where patch file belongs (this may take a while)..."
+                soft_name = scan_all_easyconfigs(patch_file)
+                if soft_name:
+                    patch_specs.append((patch_path, soft_name))
+                else:
+                    # still nothing found
+                    raise EasyBuildError("Failed to determine software name to which patch file %s relates", patch_path)
```
Comment (boegel) [inline easybuild/tools/github.py]: this whole block that determines to which software each patch file belongs should be fleshed out in a separate function, including this like this makes `_easyconfigs_pr_common` too big

File: easybuild/tools/github.py [hunk 11]
```diff
@@ -617,14 +617,53 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            for ec in file_info['ecs']:
+                if patch_file in ec['patches']:
+                    soft_name = ec.name
+                    break
+
+            if soft_name:
+                patch_specs.append((patch_path, soft_name))
+            else:
+                # fall back on scanning all eb files for patches
+                print "Matching easyconfig not found: determining where patch file belongs (this may take a while)..."
```
Comment (boegel) [inline easybuild/tools/github.py]: mention for which patch file the matching easyconfig was not found, and mention "on the first try" and that you're taking *all* available easyconfigs into account on a 2nd attempt

File: easybuild/framework/easyconfig/easyconfig.py (hunk 9 above)
Comment (Caylo) [inline easybuild/framework/easyconfig/easyconfig.py]: i needed a replacement for 'ecs' as these are not easyconfig files. for the purpose of the pr title. it can probably be scratched now

File: easybuild/tools/github.py [hunk 12]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
     return file_info, git_repo, pr_branch, diff_stat
 
 
+def scan_all_easyconfigs(patch_name):
+    """
+    Scan all easyconfigs in the robot path to determine which software a patch file belongs to
+
+    @param patch_name: name of the patch file
+    """
+    robot_path = build_option('robot_path')[0]
```
Comment (boegel) [inline easybuild/tools/github.py]: why the `[0]`? you're only considering the first entry in the robot search path? why?

File: easybuild/tools/github.py [hunk 13]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+    soft_name = ""
+    found = False
+    counter = 0
+    for (dirpath, _, filenames) in os.walk(robot_path):
+        if found:
+            break
+        for filename in [fn for fn in filenames if ".eb" in fn]:
+            rawtxt = read_file(os.path.join(dirpath, filename))
+            if "patches" in rawtxt and "TEMPLATE" not in filename:
```
Comment (boegel) [inline easybuild/tools/github.py]: why skip `TEMPLATE`, it'll never result in a match anyway?

File: easybuild/tools/github.py [hunk 14]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+    @param patch_name: name of the patch file
+    """
+    robot_path = build_option('robot_path')[0]
+    soft_name = ""
+    found = False
+    counter = 0
+    for (dirpath, _, filenames) in os.walk(robot_path):
+        if found:
+            break
+        for filename in [fn for fn in filenames if ".eb" in fn]:
```
Comment (boegel) [inline easybuild/tools/github.py]: better check the extension via `os.splitext` or via `.endswith`, and what if it's a `.yeb` easyconfig?

File: easybuild/tools/github.py [hunk 15]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            rawtxt = read_file(os.path.join(dirpath, filename))
+            if "patches" in rawtxt and "TEMPLATE" not in filename:
+                try:
+                    ec = EasyConfig(os.path.join(dirpath, filename), validate=False)
+                    ec._generate_template_values()
+                    if patch_name in ec.asdict()['patches']:
+                        soft_name = ec.asdict()['name']
+                        found = True
+                except EasyBuildError:
+                    pass
```
Comment (boegel) [inline easybuild/tools/github.py]: at least log a debug message that parsing this easyconfig failed:

```python
path = os.path.join(dirpath, filename)
...
except EasyBuildError as err:
    _log.debug("Ignoring easyconfig %s that fails to parse: %s", path, err)

File: easybuild/tools/github.py [hunk 16]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            sys.stdout.write('\r%s easyconfigs checked' % counter)
```
Comment (boegel) [inline easybuild/tools/github.py]: also mention how many easyconfigs will be checked in total (so construct a list of paths to all easyconfig files first via `os.walk`, and then simply loop over that list

File: easybuild/tools/github.py [hunk 17]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            sys.stdout.flush()
+            counter = counter+1
```
Comment (boegel) [inline easybuild/tools/github.py]: you could use `for idx, ec_path in enumerate(ec_paths)` here, and then use `idx + 1` as counter

File: easybuild/tools/github.py [hunk 18]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
     return file_info, git_repo, pr_branch, diff_stat
 
 
+def scan_all_easyconfigs(patch_name):
```
Comment (boegel) [inline easybuild/tools/github.py]: please add a test for this new function

File: easybuild/tools/github.py (hunk 13 above)
Comment (Caylo) [inline easybuild/tools/github.py]: TEMPLATE has the "patches"  parameter, so it will match, but can't be parsed as an EasyConfig in the line below
i could change the check to `patches = [.+]` to make sure it's not an empty list?

File: easybuild/tools/github.py [hunk 19]
```diff
@@ -684,6 +727,38 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+    found = False
+    counter = 0
+    for (dirpath, _, filenames) in os.walk(robot_path):
+        if found:
+            break
+        for filename in [fn for fn in filenames if ".eb" in fn]:
+            rawtxt = read_file(os.path.join(dirpath, filename))
+            if "patches" in rawtxt and "TEMPLATE" not in filename:
+                try:
+                    ec = EasyConfig(os.path.join(dirpath, filename), validate=False)
```
Comment (boegel) [inline easybuild/tools/github.py]: don't use `EasyConfig` directly, use `process_easyconfig` instead (which implements caching, so the 2nd time you'd bypass actually reparsing the easyconfig file)

File: easybuild/framework/easyconfig/easyconfig.py (hunk 9 above)
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: we shouldn't auto-guess the PR title if only patch files are being added, so yes, I would drop this

File: easybuild/tools/github.py (hunk 13 above)
Comment (boegel) [inline easybuild/tools/github.py]: OK, but then skip it with `filename != 'TEMPLATE.eb'`

File: easybuild/framework/easyconfig/easyconfig.py [hunk 20]
```diff
@@ -1365,55 +1365,96 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
+
+        target_path = os.path.join(target_dir, target_path)
+
+    else:
+        raise EasyBuildError("Subdirectory %s not found in %s", subdir, target_dir)
+
+    return target_path
+
+
+def copy_file(path, target_dir, soft_name, target_name):
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: @Caylo I'd move this to `easybuild.tools.filetools`, it's a useful function in general

File: easybuild/framework/easyconfig/easyconfig.py [hunk 21]
```diff
@@ -1365,55 +1365,96 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
+
+    target_path = det_location_for(path, target_dir, soft_name, target_name)
+    new = os.path.exists(target_path)
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: move this out of this function, this should b a generic function to copy files (also: add a docstring)

File: easybuild/framework/easyconfig/easyconfig.py [hunk 22]
```diff
@@ -1365,55 +1365,96 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
+
+    try:
+        mkdir(os.path.dirname(target_path), parents=True)
+        shutil.copy2(path, target_path)
+        _log.info("%s copied to %s", path, target_path)
+    except OSError as err:
+        raise EasyBuildError("Failed to copy %s to %s: %s", path, target_path, err)
+
+    return target_path, new
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: I don't think this should return anything, you should pass is both the source and target path (and the `new` thing is something very specific, not relevant in general)

File: easybuild/tools/github.py [hunk 23]
```diff
@@ -627,6 +627,22 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
 
     _log.debug("git status: %s", git_repo.git.status())
 
+    # seperate easyconfigs and patch files
+    ec_paths = [p for p in existing_paths if p.endswith('.eb') or p.endswith('.yeb') or not is_patch_file(p)]
```
Comment (boegel) [inline easybuild/tools/github.py]: the two `endswith` checks are useless if you also check `not is_patch_file`, since non-easyconfig files will fall through on the last check anyway

so, just use `not is_patch_file(p))` (and assume only easyconfigs are passed in)

File: easybuild/tools/github.py [hunk 24]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+    """
+    Scan all easyconfigs in the robot path to determine which software a patch file belongs to
+
+    @param patch_name: name of the patch file
+    """
+    robot_paths = build_option('robot_path')
+    soft_name = ""
+    found = False
+    for robot_path in robot_paths:
+        all_ecs = []
```
Comment (boegel) [inline easybuild/tools/github.py]: move this out of this loop, and construct a *full* set of easyconfigs first before iterating over them?

```python
all_ecs = []
for robot_path in robot_paths:
    ...

for idx, filename in enumerate(all_ecs):
    ...

File: easybuild/tools/github.py [hunk 25]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+        for (dirpath, _, filenames) in os.walk(robot_path):
+            all_ecs.extend([os.path.join(dirpath, fn) for fn in filenames])
+
+        for idx, filename in enumerate(all_ecs):
+            if found:
+                break
+            if os.path.splitext(filename)[-1] in ['.eb', '.yeb']:
```
Comment (boegel) [inline easybuild/tools/github.py]: use the `EB_FORMAT_EXTENSION` and `YEB_FORMAT_EXTENSION` constants

File: easybuild/tools/github.py [hunk 26]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                path = os.path.join(dirpath, filename)
+                rawtxt = read_file(path)
+                if "patches" in rawtxt and filename != "TEMPLATE.eb":
```
Comment (boegel) [inline easybuild/tools/github.py]: filter out `TEMPLATE.eb` when you build the `all_ecs` list

File: easybuild/tools/github.py (hunk 25 above)
Comment (boegel) [inline easybuild/tools/github.py]: and maybe also filter on the presence of `patches` already? that'll make this function a bit easier to follow

it does mean reading all files which contain 'patches' twice, but thanks to filesystem caching you'll probably not notice

File: easybuild/tools/github.py [hunk 27]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                    try:
+                        ecs = process_easyconfig(path, validate=False)
+                        for ec in ecs:
+                            ec['ec']._generate_template_values()
+                            if patch_name in ec['ec'].asdict()['patches']:
+                                soft_name = ec['ec'].asdict()['name']
+                                found = True
+                    except EasyBuildError as err:
+                        _log.debug("Ignoring easyconfig %s that fails to parse: %s", path, err)
+                        pass
```
Comment (boegel) [inline easybuild/tools/github.py]: no need for the `pass` since you have the `log.debug`

File: easybuild/tools/github.py [hunk 28]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            if found:
+                break
+            if os.path.splitext(filename)[-1] in ['.eb', '.yeb']:
+                path = os.path.join(dirpath, filename)
+                rawtxt = read_file(path)
+                if "patches" in rawtxt and filename != "TEMPLATE.eb":
+                    try:
+                        ecs = process_easyconfig(path, validate=False)
+                        for ec in ecs:
+                            ec['ec']._generate_template_values()
```
Comment (boegel) [inline easybuild/tools/github.py]: why is this needed?

File: easybuild/tools/github.py [hunk 29]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                            if patch_name in ec['ec'].asdict()['patches']:
```
Comment (boegel) [inline easybuild/tools/github.py]: just do `ec['ec']['patches']`?

File: easybuild/tools/github.py [hunk 30]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                                soft_name = ec['ec'].asdict()['name']
+                                found = True
```
Comment (boegel) [inline easybuild/tools/github.py]: add a `break` here, we're done searching

File: easybuild/tools/github.py [hunk 31]
```diff
@@ -716,6 +736,74 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                    except EasyBuildError as err:
+                        _log.debug("Ignoring easyconfig %s that fails to parse: %s", path, err)
+                        pass
+                sys.stdout.write('\r%s of %s easyconfigs checked' % (idx+1, len(all_ecs)))
```
Comment (boegel) [inline easybuild/tools/github.py]: just do `len(all_ecs)` once outside of the `for` loop, the length won't change

File: easybuild/tools/filetools.py [hunk 32]
```diff
@@ -1324,3 +1330,13 @@ def find_flexlm_license(custom_env_vars=None, lic_specs=None):
     _log.info("Found valid license specs via %s: %s", via_msg, valid_lic_specs)
 
     return (valid_lic_specs, lic_env_var)
+
+
+def copy_file(path, target_path):
+
```
Comment (boegel) [inline easybuild/tools/filetools.py]: docstring please!

File: easybuild/framework/easyconfig/easyconfig.py [hunk 33]
```diff
@@ -1365,55 +1365,85 @@ def robot_find_minimal_toolchain_of_dependency(dep, modtool, parent_tc=None):
...
-                if letter not in a_to_z:
-                    raise EasyBuildError("Don't know which letter subdir to use for %s", name)
+            target_path = det_location_for(path, target_dir, soft_name, ec_filename)
+            new = os.path.exists(target_path)
 
-                target_path = os.path.join(subdir, letter, name, ec_filename)
-                _log.debug("Target path for %s: %s", path, target_path)
+            copy_file(path, target_path)
+            file_info['paths_in_repo'].append(target_path)
+            file_info['new'].append(new)
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: no need for a `new` variable, it's only used in one place so let's do `file_info['new'].append(os.path.exists(target_path))`

File: easybuild/tools/github.py [hunk 34]
```diff
@@ -627,6 +629,22 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
     _log.debug("git status: %s", git_repo.git.status())
 
+    # seperate easyconfigs and patch files
+    ec_paths = [p for p in existing_paths if not is_patch_file(p)]
+    patch_paths = [p for p in existing_paths if p not in ec_paths]
+
+    # copy easyconfig files to right place
+    target_dir = os.path.join(git_working_dir, pr_target_repo)
+    print_msg("copying easyconfigs to %s..." % target_dir)
+    file_info = copy_easyconfigs(ec_paths, target_dir)
```
Comment (boegel) [inline easybuild/tools/github.py]: add guard `if ec_paths:` ?

File: easybuild/tools/github.py [hunk 35]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                raise EasyBuildError("Failed to determine software name to which patch file %s relates", patch_path)
+
+    return patch_specs
+
+
+def scan_all_easyconfigs(patch_name):
+    """
+    Scan all easyconfigs in the robot path to determine which software a patch file belongs to
+
+    @param patch_name: name of the patch file
```
Comment (boegel) [inline easybuild/tools/github.py]: `:param`, and also cover `:return:`

File: easybuild/tools/github.py [hunk 36]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+    """
+    robot_paths = build_option('robot_path')
+    soft_name = ""
```
Comment (boegel) [inline easybuild/tools/github.py]: use `None` as default

File: easybuild/tools/github.py [hunk 37]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            for fn in filenames:
+                if fn != 'TEMPLATE.eb':
+                    path = os.path.join(dirpath, fn)
+                    rawtxt = read_file(path)
+                    if 'patches' in rawtxt:
+                        all_ecs.append(path)
+
+    nr_of_ecs = len(all_ecs)
+    for idx, filename in enumerate(all_ecs):
+        if found:
```
Comment (boegel) [inline easybuild/tools/github.py]: drop `found`, check with `if soft_name`

File: easybuild/tools/github.py [hunk 38]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+        for (dirpath, _, filenames) in os.walk(robot_path):
+            for fn in filenames:
+                if fn != 'TEMPLATE.eb':
+                    path = os.path.join(dirpath, fn)
+                    rawtxt = read_file(path)
+                    if 'patches' in rawtxt:
+                        all_ecs.append(path)
+
+    nr_of_ecs = len(all_ecs)
+    for idx, filename in enumerate(all_ecs):
```
Comment (boegel) [inline easybuild/tools/github.py]: `s/filename/path`, since it's actually a full path

it also helps in avoiding silly mistakes to use the same variable name as above ;)

File: easybuild/tools/github.py [hunk 39]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+            if soft_name:
+                patch_specs.append((patch_path, soft_name))
+            else:
+                # still nothing found
+                raise EasyBuildError("Failed to determine software name to which patch file %s relates", patch_path)
+
+    return patch_specs
+
+
+def scan_all_easyconfigs(patch_name):
```
Comment (boegel) [inline easybuild/tools/github.py]: rename to `find_software_name_for_patch`?

File: easybuild/tools/github.py [hunk 40]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+
+    nr_of_ecs = len(all_ecs)
+    for idx, filename in enumerate(all_ecs):
+        if found:
+            break
+        rawtxt = read_file(filename)
+        try:
+            ecs = process_easyconfig(filename, validate=False)
+            for ec in ecs:
+                if patch_name in ec['ec']['patches']:
```
Comment (boegel) [inline easybuild/tools/github.py]: I just realised this is too simplistic...

The entries in `patches` can also be tuples like `('foo.patch', 2)`, where the `2` is the patch level

So, I'd create an inner function in `find_software_name_for_patch` and use that:

```python
def is_patch_for(patch_name, ec):
    """Check whether specified patch matches any patch in the provided EasyConfig instance."""
    res = False
    for patch in ec['patches']:
        if isinstance(patch, (tuple, list)):
            patch = patch[0]
        if patch == patch_name:
            res = True
            break

    return res
````

and then check with `if is_patch_for(patch_name, ec['ec'])`

File: easybuild/tools/github.py [hunk 41]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+                    soft_name = ec['ec']['name']
+                    found = True
+                    break
+        except EasyBuildError as err:
+            _log.debug("Ignoring easyconfig %s that fails to parse: %s", path, err)
```
Comment (boegel) [inline easybuild/tools/github.py]: see why you should use `path` rather than `filename`? ;)

File: easybuild/tools/github.py [hunk 42]
```diff
@@ -716,6 +738,77 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
...
+        sys.stdout.write('\r%s of %s easyconfigs checked' % (idx+1, nr_of_ecs))
+        sys.stdout.flush()
+
+    print ''
```
Comment (boegel) [inline easybuild/tools/github.py]: use `sys.stdout.write('\n')` here for consistency

File: test/framework/filetools.py [hunk 43]
```diff
@@ -672,6 +672,12 @@ def test_find_flexlm_license(self):
         del os.environ['LM_LICENSE_FILE']
         self.assertEqual(ft.find_flexlm_license(lic_specs=[None]), ([], None))
 
+    def test_is_patch_file(self):
+        """Test for is_patch_file() function."""
+        testdir = os.path.dirname(os.path.abspath(__file__))
+        self.assertFalse(ft.is_patch_file(os.path.join(testdir, 'easyconfigs', 'toy-0.0.eb')))
+        self.assertTrue(ft.is_patch_file(os.path.join(testdir, 'sandbox', 'sources', 'toy', 'toy-0.0_typo.patch')))
+
```
Comment (boegel) [inline test/framework/filetools.py]: @Caylo: please add a test for `copy_file` here too

File: easybuild/tools/github.py [hunk 44]
```diff
@@ -659,9 +659,6 @@ def _easyconfigs_pr_common(paths, start_branch=None, pr_branch=None, target_acco
             else:
                 raise EasyBuildError("Path doesn't exist or file to delete isn't found in target branch: %s", fn)
 
-    # copy edited/added files to right place
-    file_info = copy_easyconfigs(ec_paths, repo_path)
```
Comment (boegel) [inline easybuild/tools/github.py]: @Caylo: oops?

File: test/framework/github.py [hunk 45]
```diff
@@ -240,6 +242,28 @@ def test_validate_github_token(self):
...
+            'allow_modules_tool_mismatch': True,
+            'minimal_toolchains': True,
+            'use_existing_modules': True,
+            'external_modules_metadata': ConfigObj(),
+            'robot_path': [ec_path],
+            'valid_module_classes': module_classes(),
+            'validate': False,
+        })
+        self.mock_stdout(True)
+        ec = gh.scan_all_easyconfigs('toy-0.0_typo.patch')
```
Comment (boegel) [inline test/framework/github.py]: this won't work :)

File: easybuild/tools/github.py (hunk 34 above)
Comment (boegel) [inline easybuild/tools/github.py]: nvm

**Comment by boegel** [comment]
Going in, thanks @Caylo!
